Document/0x04c-Tampering-and-Reverse-Engineering.md

```diff
@@ -196,31 +197,40 @@ Even though most professional GUI-based disassemblers feature scripting faciliti
+
+In a real world situation, the functions are much more complex than the above example. The increased complexity of the functions can pose significant challenges for symbolic execution. Some of the challenges are summarised below: 
+
+- Loops and recursions in a program may lead to *infinite execution tree*
+- Multiple conditional branches or nested conditions may lead to *path explosion*
+- Complex equations generated post symbolic execution may not be solvable by SMT solvers because of their limitations
+- Program is using system calls, library calls or network events which cannot be handled by symbolic execution
+
+To overcome these challenges, typically, symbolic execution is combined with other techniques such as dynamic execution (also called concrete execution) to mitigate the path explosion problem specific to classical symbolic execution. This combination of concrete (actual) and symbolic execution is referred to as *concolic execution* (the name concolic stems from *conc*rete and symb*olic*), sometimes also called as *dynamic symbolic execution*. To visualize this, in the above example, we can obtain the value of the global variable by dynamically executing the program and feeding this information into our symbolic execution analysis. This extra information will reduce the complexity of our equations and may produce more accurate analysis results. Together with improved SMT solvers and current hardware speeds, concolic execution allows to explore paths in medium-size software modules (i.e., on the order of 10s KLOC). However, it also comes in handy for supporting de-obfuscation tasks, such as simplifying control flow graphs. For example, Jonathan Salwan and Romain Thomas have [shown how to reverse engineer VM-based software protections using Dynamic Symbolic Execution](https://triton.quarkslab.com/files/csaw2016-sos-rthomas-jsalwan.pdf "Jonathan Salwan and Romain Thomas: How Triton can help to reverse virtual machine based software protections") (i.e., using a mix of actual execution traces, simulation, and symbolic execution).
```

Comment:
Great piece of writing! Did you have other resources in here which we could use in the reference section? And can you add https://triton.quarkslab.com/files/csaw2016-sos-rthomas-jsalwan.pdf to the reference section as well please?

Reply:
I do know about few more resources, but reluctant to add them as I haven't used them. meanwhile, I will add triton.